src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs

```diff
@@ -2120,5 +2118,159 @@ private CompilationVerifier CompileAndVerify(string source, string expectedOutpu
+
+        class InstrumentationChecker
+        {
+            StringBuilder consoleExpectations = new StringBuilder();
+            Dictionary<int /* method */, MethodChecker> spanExpectations = new Dictionary<int, MethodChecker>();
+
+            public InstrumentationChecker()
+            {
+                consoleExpectations.AppendLine($"Flushing");
+            }
+
+            /// <summary>
+            /// Start recording expectations for a method.
+            /// </summary>
+            public MethodChecker M(int method, int file, string snippet)
+            {
+                consoleExpectations.AppendLine($"Method {method}");
+                consoleExpectations.AppendLine($"File {file}");
+
+                var result = new MethodChecker(consoleExpectations).T(snippet);
+                spanExpectations.Add(method, result);
+                return result;
+            }
+
+            /// <summary>
+            /// Start recording expectations for a method, but without checking snippets for this method.
+            /// </summary>
+            public MethodChecker M(int method, int file)
+            {
+                consoleExpectations.AppendLine($"Method {method}");
+                consoleExpectations.AppendLine($"File {file}");
+
+                return new MethodChecker(consoleExpectations, noSnippets: true).T();
+            }
+
+            public string ExpectedOutput { get { return consoleExpectations.ToString(); } }
+
+            /// <summary>
+            /// Verify the recorded expectations.
+            /// </summary>
+            internal void CompleteCheck(Compilation compilation, string source)
+            {
+                var peImage = compilation.EmitToArray(EmitOptions.Default.WithInstrumentationKinds(ImmutableArray.Create(InstrumentationKind.TestCoverage)));
+                var peReader = new PEReader(peImage);
+                var reader = DynamicAnalysisDataReader.TryCreateFromPE(peReader, "<DynamicAnalysisData>");
+                string[] sourceLines = source.Split('\n');
+
+                if (spanExpectations.Count == 0)
+                {
+                    Dump(reader, sourceLines);
+                    return;
+                }
+
+                foreach (int method in spanExpectations.Keys)
+                {
+                    var actualSnippets = GetActualSnippets(method, reader, sourceLines);
+                    var expectedSnippets = spanExpectations[method].SnippetExpectations;
+
+                    if (actualSnippets.Length != expectedSnippets.Length)
+                    {
+                        Assert.False(true, $"The number of spans was expected to be {expectedSnippets.Length} for the method {method}, but it actually has {actualSnippets.Length}.");
+                    }
+
+                    for (int index = 0; index < actualSnippets.Length; index++)
```

Comment:
More generally on the usability question, I've only had limited experience using the APIs and responding to the assert failures (only ported one test). But so far, my impression is that it provided the right guidance. 
This should be no problem to tweak once we get more hands-on experience.